status/status.go

```diff
@@ -0,0 +1,170 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package status contains code for determining the current state of
+// the world, including the health of a load test and its resources.
+package status
+
+import (
+	"fmt"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+
+	grpcv1 "github.com/grpc/test-infra/api/v1"
+	"github.com/grpc/test-infra/config"
+	"github.com/grpc/test-infra/optional"
+)
+
+// State reflects the observed state of a resource.
+type State string
+
+const (
+	// Pending indicates that the resource has not yet been observed as
+	// succeeding or failing.
+	Pending State = "Pending"
+
+	// Succeeded indicates that the resource has terminated successfully,
+	// marked by a zero exit code.
+	Succeeded State = "Succeeded"
+
+	// Errored indicates that the resource has terminated unsuccessfully,
+	// marked by a non-zero exit code.
+	Errored State = "Failed"
+)
+
+// StateForContainerStatus accepts the status of a container and returns a
+// ContainerState and a pointer to the integer exit code. If the container has
+// not terminated, a Pending state and nil pointer are returned.
+func StateForContainerStatus(status *corev1.ContainerStatus) (State, *int32) {
+	if terminateState := status.State.Terminated; terminateState != nil {
+		var state State = Errored
+
+		if terminateState.ExitCode == 0 {
+			state = Succeeded
+		}
+
+		return state, &terminateState.ExitCode
+	}
+
+	if waitState := status.State.Waiting; waitState != nil {
+		if strings.Compare("CrashLoopBackOff", waitState.Reason) == 0 {
+			return Errored, nil
+		}
+	}
+
+	return Pending, nil
+}
+
+// StateForPodStatus accepts the status of a pod and returns a State, as well
+// as the reason and message. The reason is a camel-case word that is machine
+// comparable. The message is a human-legible description. If the pod has not
+// terminated or it terminated successfully, the reason and message strings will
+// be empty.
+func StateForPodStatus(status *corev1.PodStatus) (state State, reason string, message string) {
+	podState := Pending
+
+	for i := range status.InitContainerStatuses {
+		initContStat := &status.InitContainerStatuses[i]
+		contState, exitCode := StateForContainerStatus(initContStat)
+
+		if contState == Errored {
+			message := fmt.Sprintf("init container %q terminated with exit code %d", initContStat.Name, *exitCode)
+			return Errored, grpcv1.InitContainerError, message
+		}
+	}
+
+	for i := range status.ContainerStatuses {
+		contStat := &status.ContainerStatuses[i]
+		contState, exitCode := StateForContainerStatus(contStat)
+
+		if contState == Errored {
+			message := fmt.Sprintf("container %q terminated with exit code %d", contStat.Name, *exitCode)
+			return Errored, grpcv1.ContainerError, message
+		}
+
+		if contState != Pending {
+			podState = contState
+		}
+	}
```

Comment:
I think it is possible to have multiple run containers in pod.Containers, this part probably changes the podState every time it sees something different from pending, say we have runtime container array with status as success, success, pending, pending, we would have success as the over all podState, right?  For our case, I think it is absolute okay, just wondering.

Reply:
@wanlin31, very fair catch! As of right now, there is only 1 container so it works. However, we want anything other than Success to bubble up, not Pending. So, that way if we encounter Success, Pending, Pending... we get Pending, not success. Will push a quick fix